### PR TITLE
Adding PHPDocs to templates

### PR DESCRIPTION
**Pull request by wow-apps**
# PR Changelog

### EmptyAuthenticator.tpl.php
`src/Resources/skeleton/authenticator/EmptyAuthenticator.tpl.php`
* Added PHPDocs for class and methods
* Added method `createAuthenticatedToken` from `Symfony\Component\Security\Guard\AuthenticatorInterface`

### EmptySecurityController.tpl.php
`src/Resources/skeleton/authenticator/EmptySecurityController.tpl.php`
* Added PHPDoc for class

### LoginFormAuthenticator.tpl.php
`src/Resources/skeleton/authenticator/LoginFormAuthenticator.tpl.php`
* Added PHPDocs for class and methods
* Removed `use TargetPathTrait;` - it's already used in extended class

### Command.tpl.php
`src/Resources/skeleton/command/Command.tpl.php`
* Added PHPDocs for class, methods and variable

### Controller.tpl.php
`src/Resources/skeleton/controller/Controller.tpl.php`
* Added PHPDocs for class and method

### Controller.tpl.php
`src/Resources/skeleton/crud/controller/Controller.tpl.php`
* Added PHPDocs for class and method

### Entity.tpl.php
`src/Resources/skeleton/doctrine/Entity.tpl.php`
* Added PHPDocs for class, variable and method

### Fixtures.tpl.php
`src/Resources/skeleton/doctrine/Fixtures.tpl.php`
* Added PHPDocs for class and method

### Repository.tpl.php
`src/Resources/skeleton/doctrine/Repository.tpl.php`
* Added PHPDocs for class and methods

### Subscriber.tpl.php
`src/Resources/skeleton/event/Subscriber.tpl.php`
* Added PHPDocs for class and methods

### Type.tpl.php
`src/Resources/skeleton/form/Type.tpl.php`
* Added PHPDocs for class and methods

### UserProvider.tpl.php
`src/Resources/skeleton/security/UserProvider.tpl.php`
* Added PHPDoc for class

### Voter.tpl.php
`src/Resources/skeleton/security/Voter.tpl.php`
* Added PHPDocs for class and methods

### Encoder.tpl.php
`src/Resources/skeleton/serializer/Encoder.tpl.php`
* Added PHPDocs for class and methods

### Functional.tpl.php
`src/Resources/skeleton/test/Functional.tpl.php`
* Added PHPDoc for class

### Unit.tpl.php
`src/Resources/skeleton/test/Unit.tpl.php`
* Added PHPDoc for class

### Extension.tpl.php
`src/Resources/skeleton/twig/Extension.tpl.php`
* Added PHPDocs for class and methods

### Constraint.tpl.php
`src/Resources/skeleton/validator/Constraint.tpl.php`
* Added PHPDocs for class and methods

### Validator.tpl.php
`src/Resources/skeleton/validator/Validator.tpl.php`
* Added PHPDocs for class and methods

### Class.tpl.php
`src/Resources/skeleton/Class.tpl.php`
* Added PHPDoc for class

### CODE REVIEW

File: src/Resources/skeleton/Class.tpl.php [hunk 1]
```diff
@@ -2,6 +2,10 @@
 
 namespace <?= $namespace; ?>;
 
+/**
+ * Class <?= $class_name ?>
+ * @package <?= $namespace; ?>
+ */
```
Comment (weaverryan) [inline src/Resources/skeleton/Class.tpl.php]: These comments don’t add any value - tools already know the class and package - so we should not add these :)

File: src/Resources/skeleton/authenticator/EmptyAuthenticator.tpl.php [hunk 2]
```diff
@@ -3,51 +3,125 @@
...
 class <?= $class_name ?> extends AbstractGuardAuthenticator
 {
+    /**
+     * Does the authenticator support the given Request?
+     * If this returns false, the authenticator will be skipped.
+     *
+     * @param Request $request
+     * @return bool
+     */
     public function supports(Request $request)
```
Comment (weaverryan) [inline src/Resources/skeleton/authenticator/EmptyAuthenticator.tpl.php]: The @param doesn’t add any value - because that arg is type-hinted. For the @return, can we add a return type instead? I actually can’t temember if that’s allowed - if a class can add a return type even though the method on the interface does not have the return type. But if it’s allowed (I think it is) we should do that instead :)

File: src/Resources/skeleton/authenticator/EmptyAuthenticator.tpl.php [hunk 3]
```diff
@@ -3,51 +3,125 @@
...
+        // TODO: implement function body
     }
 
+    /**
+     * Get the authentication credentials from the request and return them as any type.
+     *
+     * @param Request $request
+     * @return mixed Any non-null value
+     * @throws \UnexpectedValueException If null is returned
+     */
```
Comment (weaverryan) [inline src/Resources/skeleton/authenticator/EmptyAuthenticator.tpl.php]: The param and throws don’t add value

**Comment by wow-apps** [comment]
Hi @weaverryan . Sorry, I was on PHPDD conference in Dresden and have no time to finish this PR.
Yes, sure, I'll take attention on your comments and finish this in a few days.

see you again on SymfonyCon this year ;)

File: src/Resources/skeleton/doctrine/Entity.tpl.php [hunk 4]
```diff
@@ -26,7 +23,7 @@ class <?= $class_name."\n" ?>
     /**
      * @return int|null
      */
-    public function getId(): ?int
```
Comment (wow-apps) [inline src/Resources/skeleton/doctrine/Entity.tpl.php]: It's php `7.0.8` defined in composer.json
But return type `or null` come just in php 7.1

**Comment by wow-apps** [comment]
Hey @weaverryan 

I've refactor code, please make a code review before I'll fix tests (some tests depends on created templates)

Few words about last commit:

- Yes, we can add _return_ type hinting if it not defined in Interface. I've add everywhere I can;
- We can't add type hinting for arguments if they are not defined in Interface;
- I've replace `\n` everywhere in templates on `PHP_EOL` for support `new line` by not UNIX hosts
- I've remove in one or two places return type hinting `something | null` - for example: https://github.com/symfony/maker-bundle/pull/284/commits/9d93c0e0df01578b052189fd2470cda91f4c8227#r226875323 because this feature comes with php 7.1, but in composer.json for current bundle php version required is 7.0.8
- PHPDocs for all extended function were replaced on `{@inheritdoc}`
- `@package` and class name removed

have a good time while reviewing ;)

File: src/Resources/skeleton/doctrine/Entity.tpl.php (hunk 4 above)
Comment (ckrack) [reply src/Resources/skeleton/doctrine/Entity.tpl.php]: make:entity [requires PHP 7.1](https://github.com/symfony/maker-bundle/blob/master/src/Maker/MakeEntity.php#L280)

File: src/Resources/skeleton/authenticator/EmptyAuthenticator.tpl.php [hunk 5]
```diff
@@ -8,46 +8,79 @@
...
+     * {@inheritdoc}
+     */
+    public function supports(Request $request): bool
     {
-        // todo
+        // TODO: implement method
     }
 
+    /**
+     * {@inheritdoc}
```
Comment (javiereguiluz) [inline src/Resources/skeleton/authenticator/EmptyAuthenticator.tpl.php]: I dislike these `{@inheritdoc}` so much 😢  Isn't PhpStorm smart enough to do that without adding this?

File: src/Resources/skeleton/authenticator/LoginFormAuthenticator.tpl.php [hunk 6]
```diff
@@ -21,27 +23,43 @@
 
 class <?= $class_name; ?> extends AbstractFormLoginAuthenticator
 {
-    use TargetPathTrait;
-
-<?= $user_is_entity ? "    private \$entityManager;\n" : null ?>
+<?= $user_is_entity ? "    /** @var EntityManagerInterface */" . PHP_EOL . "    private \$entityManager;" . PHP_EOL . PHP_EOL : null ?>
```
Comment (javiereguiluz) [inline src/Resources/skeleton/authenticator/LoginFormAuthenticator.tpl.php]: The `@var` annotations in these properties and the `__cosntruct()` method's `@param` are not needed because constructor arguments are type-hinted and IDEs can infer the types.

File: src/Resources/skeleton/command/Command.tpl.php [hunk 7]
```diff
@@ -22,19 +26,22 @@ protected function configure()
         ;
     }
 
+    /**
+     * {@inheritdoc}
+     */
     protected function execute(InputInterface $input, OutputInterface $output)
     {
-        $io = new SymfonyStyle($input, $output);
+        $symfonyStyle = new SymfonyStyle($input, $output);
```
Comment (javiereguiluz) [inline src/Resources/skeleton/command/Command.tpl.php]: I'm not sure I like this variable name. First, we use `$io` everywhere else in Symfony. Second, this object is not really a "Symfony style" but an "input + output object that displays things using the Symfony style".

File: src/Resources/skeleton/controller/Controller.tpl.php [hunk 8]
```diff
@@ -1,26 +1,30 @@
...
 
 use Symfony\Bundle\FrameworkBundle\Controller\<?= $parent_class_name; ?>;
 use Symfony\Component\Routing\Annotation\Route;
+use Symfony\Component\HttpFoundation\<?= $twig_installed ? "Response" : "JsonResponse" ?>;
 
-class <?= $class_name; ?> extends <?= $parent_class_name; ?><?= "\n" ?>
+class <?= $class_name; ?> extends <?= $parent_class_name; ?><?= PHP_EOL ?>
 {
     /**
+     * <?= ucwords($route_name) ?> index page
```
Comment (javiereguiluz) [inline src/Resources/skeleton/controller/Controller.tpl.php]: Please let's remove this generic comment because it doesn't add much value.

File: src/Resources/skeleton/crud/controller/Controller.tpl.php [hunk 9]
```diff
@@ -13,11 +13,18 @@
...
 
 /**
+ * <?= $class_name ?> class
  * @Route("<?= $route_path ?>")
  */
-class <?= $class_name ?> extends <?= $parent_class_name; ?><?= "\n" ?>
+class <?= $class_name ?> extends <?= $parent_class_name; ?><?= PHP_EOL ?>
 {
     /**
+     * <?= ucwords($route_name) ?> index page
```
Comment (javiereguiluz) [inline src/Resources/skeleton/crud/controller/Controller.tpl.php]: In this method we can remove the generic comment, the `@param` and the `@return` because IDEs can infer all that with the arguments type hints and the return types.

Same in the other methods of this class.

File: src/Resources/skeleton/doctrine/Entity.tpl.php (hunk 4 above)
Comment (javiereguiluz) [reply src/Resources/skeleton/doctrine/Entity.tpl.php]: That's correct. Let's revert this change.

File: src/Resources/skeleton/doctrine/Repository.tpl.php [hunk 10]
```diff
@@ -1,29 +1,36 @@
...
 /**
+ * Class <?= $class_name . PHP_EOL; ?>
  * @method <?= $entity_class_name; ?>|null find($id, $lockMode = null, $lockVersion = null)
  * @method <?= $entity_class_name; ?>|null findOneBy(array $criteria, array $orderBy = null)
  * @method <?= $entity_class_name; ?>[]    findAll()
  * @method <?= $entity_class_name; ?>[]    findBy(array $criteria, array $orderBy = null, $limit = null, $offset = null)
  */
 class <?= $class_name; ?> extends ServiceEntityRepository
 {
+    /**
```
Comment (javiereguiluz) [inline src/Resources/skeleton/doctrine/Repository.tpl.php]: We can remove this entire PHPdoc block for the same reasons given for the other blocks.

File: src/Resources/skeleton/validator/Constraint.tpl.php [hunk 11]
```diff
@@ -1,10 +1,11 @@
-<?= "<?php\n" ?>
+<?= "<?php" . PHP_EOL ?>
 
-namespace <?= $namespace; ?>;
+namespace <?= $namespace ?>;
 
 use Symfony\Component\Validator\Constraint;
 
 /**
+ * Class <?= $class_name . PHP_EOL ?>
```
Comment (javiereguiluz) [inline src/Resources/skeleton/validator/Constraint.tpl.php]: Let's remove this generic comment.

File: src/Resources/skeleton/Class.tpl.php [hunk 12]
```diff
@@ -1,7 +1,7 @@
-<?= "<?php\n" ?>
+<?= "<?php" . PHP_EOL ?>
```
Comment (lyrixx) [inline src/Resources/skeleton/Class.tpl.php]: This should be reverted. We use \n everywhere, every-time

File: src/Resources/skeleton/authenticator/EmptyAuthenticator.tpl.php (hunk 5 above)
Comment (lyrixx) [reply src/Resources/skeleton/authenticator/EmptyAuthenticator.tpl.php]: I would be nice indeed (I don't use PhpStorm, and I don't care about theses inheritdoc)

File: src/Resources/skeleton/authenticator/LoginFormAuthenticator.tpl.php [hunk 13]
```diff
@@ -21,27 +23,43 @@
...
     private $router;
+
+    /** @var CsrfTokenManagerInterface */
     private $csrfTokenManager;
-<?= $user_needs_encoder ? "    private \$passwordEncoder;\n" : null ?>
 
+<?= $user_needs_encoder ? "    /** @var UserPasswordEncoderInterface */" . PHP_EOL . "    private \$passwordEncoder;" . PHP_EOL : null ?>
+
+    /**
+     * <?= $class_name; ?> constructor.
```
Comment (lyrixx) [inline src/Resources/skeleton/authenticator/LoginFormAuthenticator.tpl.php]: This should be deleted. it Does not not bind new information.

File: src/Resources/skeleton/authenticator/LoginFormAuthenticator.tpl.php [hunk 14]
```diff
@@ -21,27 +23,43 @@
...
+     *
+<?= $user_is_entity ? "     * @param EntityManagerInterface       \$entityManager" . PHP_EOL : null ?>
+     * @param RouterInterface              $router
+     * @param CsrfTokenManagerInterface    $csrfTokenManager
```
Comment (lyrixx) [inline src/Resources/skeleton/authenticator/LoginFormAuthenticator.tpl.php]: This should be deleted. it Does not not bind new information.

File: src/Resources/skeleton/crud/controller/Controller.tpl.php [hunk 15]
```diff
@@ -1,10 +1,10 @@
-<?= "<?php\n" ?>
+<?= "<?php" . PHP_EOL ?>
 
 namespace <?= $namespace ?>;
 
 use <?= $entity_full_class_name ?>;
 use <?= $form_full_class_name ?>;
-<?php if (isset($repository_full_class_name)): ?>
+<?php if (!empty($repository_full_class_name)): ?>
```
Comment (lyrixx) [inline src/Resources/skeleton/crud/controller/Controller.tpl.php]: Why this change?

**Comment by weaverryan** [comment]
I'm going to close this. It needs to be updated, and while I think some changes are good, I think some changes are not needed. A smaller PR with individual changes would work really well I think :).

Cheers!
